tapeserver/cta-taped.man
.\" The CERN Tape Archive (CTA) project
.\" Copyright (C) 2015  CERN
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH CTA-TAPED "8cta" "$Date: 2014/03/24 14:44:00 $" CTA "CTA"
.SH NAME
cta-taped \- Tape server daemon
.SH SYNOPSIS
.BI "cta-taped [OPTIONS]"

.SH DESCRIPTION
\fBcta-taped\fP is the daemon responsible for controlling one or more tape
drives.
.P
When \fBcta-taped\fP is executed it immediately forks with the parent
terminating and the child running in the background.  If required the
\fB\-f, \-\-foreground\fP option can be used to stop \fBcta-taped\fP from
forking and keep the parent process in the foreground.

.SH TAPE LIBRARY SUPPORT

CTA supports two types of tape libraries, SCSI compatible ones and Oracle
Automatic Cartridge System (ACS) ones.  The cta-taped daemon requires a tape
library daemon to be installed and run on the same tape server as itself.
There are two types of CTA tape library daemons, one for SCSI compatible tape
libraries and one for ACS tape libraries.

A SCSI compatible tape library requires the following daemon:

\fBrmcd\fP

The installation RPM for the \fBrmcd\fP daemon and its manual page is:

\fBcta-rmc-server-0.0-XX.slc6.x86_64\fP

The \fBcta-rmcd\fP daemon can be started, stopped and its status queried using
the usual service commands:

\fBservice cta-rmcd start\fP
.br
\fBservice cta-rmcd status\fP
.br
\fBservice cta-rmcd stop\fP

An ACS tape library requires the following daemon:

\fBcta-acsd\fP

The installation RPM for the \fBcta-acsd\fP daemon and its manual page is:

\fBcta-acs-server-0.0.slc6.x86_64\fP

The \fBcta-acsd\fP daemon can be started, stopped and its status queried using
the usual service commands:

\fBservice cta-acsd start\fP
.br
\fBservice cta-acsd status\fP
.br
\fBservice cta-acsd stop\fP

Even though there is a separate daemon for each of the supported tape library
types, there is a single generic set of command-line tools for mounting and
un-mounting tapes:

\fBcta-tape-mediachanger-mount\fP
.br
\fBcta-tape-mediachanger-dismount\fP

The installation RPM for these command-line tools and their manual pages is:

\fBcta-tape-developer-tools-2.1.15-XX.slc6.x86_64\fP

.SH OPTIONS
.TP
\fB\-f, \-\-foreground
Remain in the foreground.
.TP
\fB\-h, \-\-help
Prints the usage message.
.TP
\fB\-c, \-\-config <config-file>
Set the location of the CTA configuration file (cta.conf).  The default location is /etc/cta/cta.conf.

.SH CTA CONFIGURATION PARAMETERS
The cta-taped daemon reads and uses the following CTA configuration
parameters which are specified within the CTA configuration file (the
default location is /etc/cta/cta.conf).

.TP
\fBRMC HOST
The host on which the rmcd daemon is running.

.TP
\fBRMC MAXRQSTATTEMPTS
Maximum number of attempts a retriable RMC request should be issued.

.TP
\fBTapeServer BlkMoveTimeout
The maximum time in seconds the data-transfer session of tapeserverd can
cease to move data blocks

.TP
\fBTapeServer BufSize
Size of a memory buffer in the data-transfer cache in bytes (default is 5
Mebibytes).

.TP
\fBTapeServer BulkRequestMigrationMaxBytes
When the tapeserverd daemon requests the tapegatewayd daemon for a set of
files to migrate to tape, this parameter defines the maximum number of bytes
the set of files should represent.

.TP
\fBTapeServer BulkRequestMigrationMaxFiles
When the tapeserverd daemon requests the tapegatewayd daemon for a set of
files to migrate to tape, this parameter defines the maximum number of files
the set may contain.

.TP
\fBTapeServer BulkRequestRecallMaxBytes
When the tapeserverd daemon requests the tapegatewayd daemon for a set of
files to recall from tape, this parameter defines the maximum number of bytes
the set of files should represent.

.TP
\fBTapeServer BulkRequestRecallMaxFiles
When the tapeserverd daemon requests the tapegatewayd daemon for a set of
files to recall from tape, this parameter defines the maximum number of files
the set may contain.

.TP
\fBTapeServer MaxBytesBeforeFlush
The value of this parameter defines the maximum number of bytes to be written
to tape before a flush to tape (synchronised tape-mark).  Please note that a
flush occurs on a file boundary therefore more bytes will normally be written
to tape before the actual flush occurs.

.TP
\fBTapeServer MaxFilesBeforeFlush
The value of this parameter defines the maximum number of files to be written
to tape before a flush to tape (synchronised or non-immediate tape-mark).

.TP
\fBTapeServer MountTimeout
The maximum time in seconds that the data-transfer session can take to mount a
tape.

.TP
\fBTapeServer NbBufs
Number of memory buffers in the data-transfer cache.

.TP
\fBTapeServer NbDiskThreads
The number of disk I/O threads.

.TP
\fBTapeServer RemoteFileProtocol
The protocol to be used when transfering files to and from disk servers.
Possible values are RFIO or XROOT.  The value is not case sensitive.

.TP
\fBXROOT PrivateKey
The file hosting the Xroot private key, needed to sign Xroot URLs.

.SH FILES
.TP
.B /etc/cta/TPCONFIG
The tape drive configuration file.
.TP
.B /etc/cta/cta.conf
The CTA configuration file.
.TP
.B /var/log/cta/cta-tapeserverd.log
The tape tapeserver log file.

.SH TRADEMARKS
Oracle is a registered trademark of Oracle Corporation and/or its affiliates.

.SH AUTHOR
\fBCTA\fP Team


tapeserver/cta-taped.cpp
/*
 * The CERN Tape Archive (CTA) project
 * Copyright (C) 2015  CERN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "common/Configuration.hpp"
#include "common/log/StdoutLogger.hpp"
#include "common/log/SyslogLogger.hpp"
#include "common/processCap/ProcessCap.hpp"
#include "tapeserver/daemon/CommandLineParams.hpp"
#include "tapeserver/daemon/GlobalConfiguration.hpp"
#include "tapeserver/daemon/TpconfigLines.hpp"
#include "tapeserver/daemon/TapeDaemon.hpp"

#include "version.h"

#include <google/protobuf/stubs/common.h>
#include <memory>
#include <sstream>
#include <string>
#include <iostream>

namespace cta { namespace taped {

//------------------------------------------------------------------------------
// exceptionThrowingMain
//
// The main() function delegates the bulk of its implementation to this
// exception throwing version.
//
// @param argc The number of command-line arguments.
// @param argv The command-line arguments.
// @param log The logging system.
//------------------------------------------------------------------------------
static int exceptionThrowingMain(const int argc, char **const argv,
  cta::log::Logger &log);

//------------------------------------------------------------------------------
// The help string
//------------------------------------------------------------------------------
std::string gHelpString = 
    "Usage: cta-taped [options]\n"
    "\n"
    "where options can be:\n"
    "\n"
    "\t--foreground            or -f         \tRemain in the Foreground\n"
    "\t--stdout                or -s         \tPrint logs to standard output. Required --foreground\n"
    "\t--config <config-file>  or -c         \tConfiguration file\n"
    "\t--help                  or -h         \tPrint this help and exit\n"
    "\n"
    "Comments to: Castor.Support@cern.ch\n";

//------------------------------------------------------------------------------
// Logs the start of the daemon.
//------------------------------------------------------------------------------
static void logStartOfDaemon(cta::log::Logger &log, const int argc,
  const char *const *const argv);

//------------------------------------------------------------------------------
// Creates a string that contains the specified command-line arguments
// separated by single spaces.
//
// @param argc The number of command-line arguments.
// @param argv The array of command-line arguments.
//------------------------------------------------------------------------------
static std::string argvToString(const int argc, const char *const *const argv);

////------------------------------------------------------------------------------
//// Writes the specified TPCONFIG lines to the specified logging system.
////
//// @param log The logging system.
//// @param lines The lines parsed from /etc/cta/TPCONFIG.
////------------------------------------------------------------------------------
//static void logTpconfigLines(cta::log::Logger &log,
//  const cta::tape::daemon::TpconfigLines &lines);
//
////------------------------------------------------------------------------------
//// Writes the specified TPCONFIG lines to the logging system.
////
//// @param log The logging system.
//// @param line The line parsed from /etc/cta/TPCONFIG.
////------------------------------------------------------------------------------
//static void logTpconfigLine(cta::log::Logger &log,
//  const cta::tape::daemon::TpconfigLine &line);

//------------------------------------------------------------------------------
// exceptionThrowingMain
//------------------------------------------------------------------------------
static int exceptionThrowingMain(const int argc, char **const argv,
  cta::log::Logger &log) {
  using namespace cta::tape::daemon;

  logStartOfDaemon(log, argc, argv);

  // Parse /etc/cta/cta.conf and /etc/cta/TPCONFIG for global parameters
  const GlobalConfiguration globalConfig =
    GlobalConfiguration::createFromCtaConf(log);

  // Create the object providing utilities for working with UNIX capabilities
  cta::server::ProcessCap capUtils;

  // Create the main tapeserverd object
  cta::tape::daemon::TapeDaemon daemon(
    argc,
    argv,
    std::cout,
    std::cerr,
    log,
    globalConfig,
    capUtils);

  // Run the tapeserverd daemon
  return daemon.main();
}

//------------------------------------------------------------------------------
// logStartOfDaemon
//------------------------------------------------------------------------------
static void logStartOfDaemon(cta::log::Logger &log, const int argc,
  const char *const *const argv) {
  using namespace cta;

  const std::string concatenatedArgs = argvToString(argc, argv);
  std::list<log::Param> params = {
    log::Param("version", CTA_VERSION),
    log::Param("argv", concatenatedArgs)};
  log(log::INFO, "tapeserverd started", params);
}

//------------------------------------------------------------------------------
// argvToString
//------------------------------------------------------------------------------
static std::string argvToString(const int argc, const char *const *const argv) {
  std::string str;

  for(int i=0; i < argc; i++) {
    if(i != 0) {
      str += " ";
    }

    str += argv[i];
  }
  return str;
}

////------------------------------------------------------------------------------
//// logTpconfigLines
////------------------------------------------------------------------------------
//static void logTpconfigLines(cta::log::Logger &log,
//  const cta::tape::daemon::TpconfigLines &lines) {
//  using namespace cta::tape::daemon;
//
//  for(TpconfigLines::const_iterator itor = lines.begin();
//    itor != lines.end(); itor++) {
//    logTpconfigLine(log, *itor);
//  }
//}
//
////------------------------------------------------------------------------------
//// logTpconfigLine
////------------------------------------------------------------------------------
//static void logTpconfigLine(cta::log::Logger &log,
//  const cta::tape::daemon::TpconfigLine &line) {
//  std::list<cta::log::Param> params = {
//    cta::log::Param("unitName", line.unitName),
//    cta::log::Param("logicalLibrary", line.logicalLibrary),
//    cta::log::Param("devFilename", line.devFilename),
//    cta::log::Param("librarySlot", line.librarySlot)};
//  log(log::INFO, "TPCONFIG line", params);
//}

}} // namespace cta::taped

//------------------------------------------------------------------------------
// main
//------------------------------------------------------------------------------
int main(const int argc, char **const argv) {
  using namespace cta;
  
  // Interpret the command line
  std::unique_ptr<cta::daemon::CommandLineParams> commandLine;
  try {
    commandLine.reset(new cta::daemon::CommandLineParams(argc, argv));
  } catch (exception::Exception &ex) {
    std::cerr <<
      "Failed to interpret the command line parameters: " <<
      ex.getMessage().str() << std::endl;
    return EXIT_FAILURE;
  }
  
  if(commandLine->helpRequested) {
    std::cout << cta::taped::gHelpString << std::endl;
    return EXIT_SUCCESS;
  }

  // Try to instantiate the logging system API
  std::unique_ptr<log::Logger> logPtr;
  try {
    if (commandLine->logToStdout) {
      logPtr.reset(new log::StdoutLogger("cta-taped"));
    } else {
      logPtr.reset(new log::SyslogLogger(log::SOCKET_NAME, "cta-taped",
        log::DEBUG));
    }
  } catch(exception::Exception &ex) {
    std::cerr <<
      "Failed to instantiate object representing CTA logging system: " <<
      ex.getMessage().str() << std::endl;
    return EXIT_FAILURE;
  }
  log::Logger &log = *logPtr;

  int programRc = EXIT_FAILURE; // Default return code when receiving an exception.
  try {
    programRc = cta::taped::exceptionThrowingMain(argc, argv, log);
  } catch(exception::Exception &ex) {
    std::list<log::Param> params = {
      log::Param("message", ex.getMessage().str())};
    log(log::ERR, "Caught an unexpected CTA exception", params);
  } catch(std::exception &se) {
    std::list<log::Param> params = {log::Param("what", se.what())};
    log(log::ERR, "Caught an unexpected standard exception", params);
  } catch(...) {
    log(log::ERR, "Caught an unexpected and unknown exception");
  }

  google::protobuf::ShutdownProtobufLibrary();
  return programRc;
}
